src/providers/oauth-provider.js

```diff
@@ -0,0 +1,190 @@
+    }
+
+    return config;
+  };
+
+  /**
+   * OAuth service.
+   *
+   * @ngInject
+   */
+
+  this.$get = function($http, OAuthToken) {
+    class OAuth {
+
+      /**
+       * Check if `OAuthProvider` is configured.
+       */
+
+      constructor() {
+        if (!config) {
+          throw new Error('`OAuthProvider` must be configured first.');
+        }
+      }
+
+      /**
+       * Verifies if the `user` is authenticated or not based on the `token`
+       * cookie.
+       *
+       * @return {boolean}
+       */
+
+      isAuthenticated() {
+        return !!OAuthToken.token;
+      }
+
+      /**
+       * Retrieves the `access_token` and stores the `response.data` on cookies
+       * using the `OAuthToken`.
+       *
+       * @param {object} user - Object with `username` and `password` properties.
+       * @param {object} config - Optional configuration object sent to `POST`.
+       * @return {promise} A response promise.
+       */
+
+      getAccessToken(user, options) {
+        // Check if `user` has required properties.
+        if (!user || !user.username || !user.password) {
+          throw new Error('`user` must be an object with `username` and `password` properties.');
+        }
+
+        var data = queryString.stringify({
+          client_id: config.clientId,
+          client_secret: config.clientSecret,
+          grant_type: 'password',
+          username: user.username,
+          password: user.password
+        });
+
+        options = angular.extend({
+          headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+        }, options);
+
+        return $http.post(`${config.baseUrl}${config.grantPath}`, data, options).then((response) => {
+          OAuthToken.token = response.data;
+
+          return response;
+        });
+      }
+
+      /**
+       * Retrieves the `refresh_token` and stores the `response.data` on cookies
+       * using the `OAuthToken`.
+       *
+       * @return {promise} A response promise.
+       */
+
+      getRefreshToken() {
+        var data = queryString.stringify({
+          client_id: config.clientId,
+          client_secret: config.clientSecret,
+          grant_type: 'refresh_token',
+          refresh_token: OAuthToken.refreshToken,
+        });
+
+        var options = {
+          headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
+        };
+
+        return $http.post(`${config.baseUrl}${config.grantPath}`, data, options).then((response) => {
+          OAuthToken.token = response.data;
+
+          return response;
+        });
+      }
+
+      /**
+       * Revokes the `token` and removes the stored `token` from cookies
+       * using the `OAuthToken`.
+       *
+       * @return {promise} A response promise.
+       */
+
+      revokeToken() {
+        var data = queryString.stringify({
+          token: OAuthToken.refreshToken ? OAuthToken.refreshToken : OAuthToken.accessToken
```

Comment:
"The authorization server MAY revoke the old refresh token after issuing a new refresh token to the client". So, there's no need to pass it as an option, just revoke the latest `refresh_token` or `access_token` issued.